bazel/utils/diff_test.bzl

```diff
@@ -1,17 +1,72 @@
-"""A set of rules to compare generated output against an expected data file.
+"""A set of rules to compare generated output against an expected data file."""
 
-Example of use:
+def _zipdiff_test_impl(ctx):
+    # Note: in python format strings, {{ and }} render as { and }.
+    script = """
+        err=0
+        act="{actual}"
+        exp="{expected}"
+        if [[ ! -e "${{act}}" ]]; then echo "Missing file: ${{act}}"; exit 1; fi
+        if [[ ! -e "${{exp}}" ]]; then echo "Missing file: ${{exp}}"; exit 1; fi
+        actdir="$(mktemp -d -p ${{TEST_TMPDIR}})"
+        unzip "${{act}}" -d "${{actdir}}"
+        expdir="$(mktemp -d -p ${{TEST_TMPDIR}})"
+        unzip "${{exp}}" -d "${{expdir}}"
+        diff -r -u "${{actdir}}" "${{expdir}}"; RC=$?
+        while [ "${{1:-}}" != "" ]; do
+          if [[ ${{RC}} -ne 0 ]] && [[ "$1" == "--update_goldens" ]]; then
+            b="$(readlink -f "${{exp}}")"
+            echo "Updating ${{b}}"
+            cp -vf "${{act}}" "${{b}}"
+          fi
+          shift
+        done
+        exit ${{RC}}
+        """.format(
+        actual = ctx.files.actual[0].short_path,
+        expected = ctx.files.expected[0].short_path,
+    )
+    ctx.actions.write(
+        output = ctx.outputs.executable,
+        content = script,
+    )
+    runfiles = ctx.runfiles(files = ctx.files.actual + ctx.files.expected)
+    return [DefaultInfo(runfiles = runfiles)]
 
-  genrule(
-     name = "foobar.txt-gen",
+zipdiff_test = rule(
+    doc = """
+      A test that compares the contents of two zip files to ensure the contents
+      are identical.
 
+      Typically, this would be used to compare the contents of a generated file
+      against an expected data file.
 
-"""
+      A quick way to update expected data files:
+
+          blaze run :some_diff_test -- --update_goldens
+
+    """,
+    implementation = _zipdiff_test_impl,
+    attrs = {
+        "expected": attr.label(
+            doc = "A label indicating the file containing the expected data.",
+            allow_files = True,
+        ),
+        "actual": attr.label(
+            doc = "A label indicating the file containing the actual data to check.",
+            allow_files = True,
+        ),
+    },
+    test = True,
+)
 
 def _diff_test_impl(ctx):
-    if len(ctx.files.actual) > 1:
-        fail("`actual` must specify a single file.")
-    actual = ctx.files.actual[0].short_path
+    if ctx.attr.output_within_actual:
```

Comment:
I didn't integrate this functionality when I pushed diff_test public because it's a bit of a mess.

A more bazel-ic implementation would add an optional "deps" attribute to diff_test, that can be used to add a dependency with undeclared outputs.

Reply:
I'm personally more concerned about the divergence of the implementations than the cleanness of it :/
What you have works, and if there are improvements or fixes, ... at least we'll benefit from it everywhere?
I can withdraw the PR, or parts of it if you feel strongly, but I do think there's a benefit from sharing this code.

Reply:
I hear you.  OK, let's check in as-is and I'll do a two-stage cleanup later.